src/covid_model_seiir_pipeline/io/keys.py

```diff
@@ -0,0 +1,124 @@
+"""Primitives for representing logical datasets and groups for I/O.
+
+
+"""
+
+from pathlib import Path
+from typing import NamedTuple, Optional, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from .data_roots import DataRoot
+
+
+class __LeafTemplates(NamedTuple):
+    """Templates for leaf nodes in the on disk representation of datasets.
+
+    The structure of a leaf node depends on the on disk format of the data.
+    For csvs, a leaf node might be a file.  For hdfs, on the other hand,
+    it is an actual node enclosed in a hierarchy in the hdf file.
+
+    """
+    DRAW_TEMPLATE: str = 'draw_{draw_id}'
+    MEASURE_TEMPLATE: str = '{measure}'
+    COV_SCENARIO_TEMPLATE: str = '{covariate_scenario}_scenario'
+    COV_INFO_TEMPLATE: str = '{info_type}_info'
+
+
+LEAF_TEMPLATES = __LeafTemplates()
+
+
+class __PathTemplates(NamedTuple):
```

Comment:
This is maybe better interpreted as a prefix template.  Things get parsed as 
``root/path_template/dataset_key.data_type/leaf_template``

Reply:
path template always resolves to nested directories (in the current implementation).  dataset_key.datatype will turn into a file for hdf/zip and a terminal subdirectory for csv.  leaf_template is a node in hdf/zip and a file in csv.

Reply:
Settled on prefix templates